File: cpp_10-99/16.cc
/**
 * C++
 * leetcode 16
 * url: https://leetcode-cn.com/problems/3sum-closest/
 * 先排序
 * 用i遍历作第一个数字
 * 然后j,k双指针
 * 使用ret记录需要返回的值，用decalage记录 ret 与target的绝对值，并且使用decalage作为更改ret的依据
 */

class Solution {
public:
    int abso(int a, int b){
        if(a>b) return a-b;
        return b-a;
    }

    int threeSumClosest(vector<int>& nums, int target) {
        int size = nums.size();
        sort( nums.begin(), nums.end());
        if(nums[0] + nums[1] +nums[2] >target)
            return nums[0] + nums[1] +nums[2];
        if(nums[size-1] +nums[size-2] +nums[size-3] <target)
            return nums[size-1] +nums[size-2] +nums[size-3];
        int i, j, k, i_val, j_val, k_val, temp;
        int ret, decalage = INT_MAX;
        for( int i =0; i<size-2; i++)
        {
            i_val = nums[i];
            j = i+1;
            j_val = nums[j];
            k = size-1;
            k_val = nums[k];
            while(j<k)
            {   
                temp = i_val+j_val+k_val;
                if(abso(temp, target) < decalage)
                {
                    ret = temp;
                    decalage = abso(temp, target);
                }
                if(temp > target)
                {   // 刷k
                    while(nums[k] ==k_val && j<k)
                        k--;
                    k_val = nums[k];
                }
                else if(temp < target)
                {   // 刷j
                    while(nums[j] ==j_val && j<k)
                        j++;
                    j_val = nums[j];
                }else{
                    return target;
                }
            }
        }   // end i loop
        return ret;
    }
};

File: Go/Go_10-99/16.go
// Go
// leetcode 16
// https://leetcode.cn/problems/3sum-closest/

// dp
func threeSumClosest(nums []int, target int) int {
	f := make([][]bool, 3)
	for i:=0; i<3; i++ {
		f[i] = make([]bool, 6001)
	}
	for _, x := range nums {
		for i:=1000; i<=5000; i++ {
			if f[1][i] {
				f[2][i+x] = true
			}
		}

		for i:= 2000; i<=4000; i++ {
			if f[0][i] {
				f[1][i+x] = true
			}
		}
		
		f[0][x+3000] =true
	}

	if target > 3000 {
		for i:=6000; i >= 0; i-- {
			if f[2][i] {
				return i-3000
			}
		}
	}

	if target < -3000 {
		for i:=0; i <= 6000; i++ {
			if f[2][i] {
				return i-3000
			}
		}
	}
	u := target + 3000
	good := func(x int) bool {
		if x < 0 || x > 6000 {
			return false
		}
		return f[2][x]
	}

	for i:=0; i<=6000; i++ {
		if good(u+i) {
			return u+i -3000
		}
		if good(u-i) {
			return u-i -3000
		}
	}

	return 10000 // never
}


// // 双指针
// func threeSumClosest(nums []int, target int) int {
// 	sort.Ints(nums)

// 	res := nums[0]+nums[1]+nums[2]
// 	updateRes := func(x int) {
// 		diff := target - x
// 		if diff < 0 {
// 			diff = -diff
// 		}
// 		cur := target - res 
// 		if cur < 0 {
// 			cur = -cur
// 		}
// 		if diff < cur {
// 			res = x
// 		}
// 	}

// 	n := len(nums)
// 	for i:=0; i<n; i++ {
// 		firstNum := nums[i]
// 		l, r := i+1, n-1
// 		for l<r {
// 			cur := firstNum + nums[l] + nums[r]
// 			updateRes(cur)
// 			if cur > target {
// 				r--
// 			} else if cur <  target{
// 				l++
// 			} else {
// 				return target
// 			}
// 		}
// 	}
// 	return res
// }